### PR TITLE
Read CARFIN and create LGR object

### PR DESCRIPTION
**Pull request by goncalvesmachadoc**

### CODE REVIEW

File: src/opm/input/eclipse/EclipseState/Grid/Carfin.cpp [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        this->m_active_index_list.clear();
+        this->m_global_index_list.clear();
+
+        const auto lgrdims = GridDims(this->m_dims[0], this->m_dims[1], this->m_dims[2]);
+        const auto ncells = lgrdims.getCartesianSize();
+
+        for (auto data_index = 0*ncells; data_index != ncells; ++data_index) {
+            const auto lgrIJK = lgrdims.getIJK(data_index);
+            const auto global_index = this->m_globalGridDims_
+                .getGlobalIndex(lgrIJK[0]/this->m_dims[0]*(this->m_end_offset[0]-this->m_offset[0]+1) + this->m_offset[0],
```
Comment (blattms) [inline src/opm/input/eclipse/EclipseState/Grid/Carfin.cpp:215]: Let's skip the multiplication with `this->m_end_offset[i]-this->m_offset[i]+1`. Should be 1 all the time

File: opm/input/eclipse/EclipseState/Grid/Carfin.hpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+            cell_index(std::size_t g,std::size_t a, std::size_t d)
+                : global_index(g)
+                , active_index(a)
+                , data_index(d)
+            {}
+
+            cell_index(std::size_t g, std::size_t d)
+                : global_index(g)
+                , active_index(g)
```
Comment (blattms) [inline opm/input/eclipse/EclipseState/Grid/Carfin.hpp:57]: I think we should use some marker here, to somehow indicate that this is inactive, e.g.
```
#include<numeric_limits>
....
, active_index(std::numeric_limits<std::size_t>::max())
```

File: opm/input/eclipse/EclipseState/Grid/Carfin.hpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        GridDims m_globalGridDims_{};
+        IsActive m_globalIsActive_{};
+        ActiveIdx m_globalActiveIdx_{};
+
+        std::array<std::size_t, 3> m_dims{};
+        std::array<std::size_t, 3> m_offset{};
+        std::array<std::size_t, 3> m_end_offset{};
+
+        std::vector<cell_index> m_active_index_list;
+        std::vector<cell_index> m_global_index_list;
```
Comment (blattms) [inline opm/input/eclipse/EclipseState/Grid/Carfin.hpp:110]: Not really clear from the name what the difference is. Seems like one is the complete list (inactive and active, the triplet (global_in_global_grid, global_in_global_grid, lgr_global) and the other just the active ones, the triplet (global_in_global_grid, active_in_global_grid, lgr_global).

I assume that we should use m_active_index_list for setting up the LGR in the grid. What is the intended purpose of the other ones?

**Review by blattms — COMMENTED** [review]
Thanks a lot.

Not a complete review. Just a few questions and comments, while trying to figure what information the grid will get about an LGR.

Would be cool if you could add some comments here in there to help me and @aritorto a bit.

File: opm/input/eclipse/EclipseState/Grid/Carfin.hpp (hunk 2 above)
Comment (bska) [reply opm/input/eclipse/EclipseState/Grid/Carfin.hpp:57]: > I think we should use some marker here, to somehow indicate that this is inactive, e.g.

I suppose that's debatable.  This particular constructor is used here to initialise the "global" index list and it's not clear to me that marking the cell as inactive is the right thing to do.

**Review by bska — COMMENTED** [review]
Just a couple of remarks, mostly in response to @blattms' earlier review.  We should probably get a more comprehensive test suite soon though.

**Comment by goncalvesmachadoc** [comment]
jenkins build this please

**Comment by bska** [comment]
> jenkins build this please

Thanks for the updates, @goncalvesmachadoc.  At this moment, please be aware that all build checks fail because of an external ALUGrid problem on the CI system.  We're essentially blocked by OPM/opm-simulators#4269 which is in review.

**Comment by bska** [comment]
jenkins build this please

**Comment by goncalvesmachadoc** [comment]
jenkins build this please

**Comment by goncalvesmachadoc** [comment]
jenkins build this please

**Comment by goncalvesmachadoc** [comment]
jenkins build this please

**Comment by goncalvesmachadoc** [comment]
jenkins build this please

File: opm/input/eclipse/EclipseState/Grid/Carfin.hpp (hunk 3 above)
Comment (blattms) [reply opm/input/eclipse/EclipseState/Grid/Carfin.hpp:107]: In the mean time I might have understood what this is, by reading the code constructing it.

Is the value at index i inherited from the cell that was refined to produce cell i of the LGR? Hence it is the active and global index of those cell.

Maybe we should put add a comment about this?

**Comment by goncalvesmachadoc** [comment]
jenkins build this please

**Comment by bska** [comment]
Hi @goncalvesmachadoc,

I think this looks good for the most part.  I'll do a more thorough review later.  I notice that one of the functions in the new unit test, specifically [`makeState()`](https://github.com/OPM/opm-common/blob/7c29f6c2a526cc9cdd625229fdc3d62fefee35d9/tests/parser/integration/CarfinTest.cpp#L48-L50), is [unused](https://ci.opm-project.org/job/opm-common-PR-builder/5614/gcc/new/).  Is that intentional?  If not, maybe we can remove that function?

**Comment by goncalvesmachadoc** [comment]
> Hi @goncalvesmachadoc,
> 
> I think this looks good for the most part. I'll do a more thorough review later. I notice that one of the functions in the new unit test, specifically [`makeState()`](https://github.com/OPM/opm-common/blob/7c29f6c2a526cc9cdd625229fdc3d62fefee35d9/tests/parser/integration/CarfinTest.cpp#L48-L50), is [unused](https://ci.opm-project.org/job/opm-common-PR-builder/5614/gcc/new/). Is that intentional? If not, maybe we can remove that function?

Deleted. I may need need later, but will see to that in the following-up PR.

**Comment by goncalvesmachadoc** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks a lot for the updates, @goncalvesmachadoc , and apologies about the delayed review.  I'm happy with the current version and I'll merge into the master branch.

**Comment by goncalvesmachadoc** [comment]
> Thanks a lot for the updates, @goncalvesmachadoc , and apologies about the delayed review. I'm happy with the current version and I'll merge into the master branch.

Thank you for the guidance. :)

**Pull request merged**